rust/src/enip/detect.rs

```diff
@@ -0,0 +1,334 @@
+) -> std::os::raw::c_int {
+    let tx = cast_pointer!(tx, EnipTransaction);
+    return enip_tx_is_cmd(tx, direction.into(), value);
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct DetectCipServiceData {
+    pub service: u8,
+    pub class: Option<u16>,
+    pub attribute: Option<u16>,
+}
+
+fn enip_parse_cip_service(i: &str) -> IResult<&str, DetectCipServiceData> {
+    let (i, _) = space0(i)?;
+    let (i, service) = verify(map_opt(digit1, |s: &str| s.parse::<u8>().ok()), |&v| {
+        v < 0x80
+    })(i)?;
+    let mut class = None;
+    let mut attribute = None;
+    let (i, _) = space0(i)?;
+    let (i, comma) = opt(char(','))(i)?;
+    let mut input = i;
+    if comma.is_some() {
+        let (i, _) = space0(i)?;
+        let (i, class1) = map_opt(digit1, |s: &str| s.parse::<u16>().ok())(i)?;
+        class = Some(class1);
+        let (i, _) = space0(i)?;
+        let (i, comma) = opt(char(','))(i)?;
+        input = i;
+        if comma.is_some() {
+            let (i, _) = space0(i)?;
+            let (i, negation) = opt(char('!'))(i)?;
+            let (i, attr1) = map_opt(digit1, |s: &str| s.parse::<u16>().ok())(i)?;
+            if negation.is_none() {
+                attribute = Some(attr1);
+            }
+            input = i;
+        }
+    }
+    let (i, _) = space0(input)?;
+    if !i.is_empty() {
+        return Err(nom7::Err::Error(make_error(i, ErrorKind::NonEmpty)));
+    }
+    return Ok((
+        i,
+        DetectCipServiceData {
+            service,
+            class,
+            attribute,
+        },
+    ));
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_enip_parse_cip_service(
+    raw: *const std::os::raw::c_char,
+) -> *mut c_void {
+    let raw2: &CStr = CStr::from_ptr(raw); //unsafe
+    if let Ok(s) = raw2.to_str() {
+        if let Ok((_, ctx)) = enip_parse_cip_service(s) {
+            let boxed = Box::new(ctx);
+            return Box::into_raw(boxed) as *mut _;
+        }
+    }
+    return std::ptr::null_mut();
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_enip_cip_service_free(ctx: *mut c_void) {
+    // Just unbox...
+    std::mem::drop(Box::from_raw(ctx));
+}
+
+fn enip_cip_has_attribute(cipdir: &CipDir, class: u16) -> std::os::raw::c_int {
+    if let CipDir::Request(req) = cipdir {
+        for seg in req.path.iter() {
+            match seg.segment_type {
+                ENIP_CIP_PATH_ATTR_8BIT | ENIP_CIP_PATH_ATTR_16BIT => {
+                    if seg.value == class {
+                        return 1;
+                    }
+                }
+                _ => {}
+            }
+        }
+    }
+    return 0;
+}
+
+fn enip_cip_has_class(cipdir: &CipDir, class: u16) -> bool {
+    if let CipDir::Request(req) = cipdir {
+        for seg in req.path.iter() {
+            match seg.segment_type {
+                ENIP_CIP_PATH_CLASS_8BIT | ENIP_CIP_PATH_CLASS_16BIT => {
+                    if seg.value == class {
+                        return true;
```

Comment:
Can this conditional be put on the match arm? https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html#extra-conditionals-with-match-guards?

Reply:
Working locally, will try CI with MSRV next PR ;-)